src/timely-util/src/builder_async.rs

```diff
@@ -0,0 +1,361 @@
+    ) -> (
+        OutputWrapper<G::Timestamp, D, TeeCore<G::Timestamp, D>>,
+        StreamCore<G, D>,
+    ) {
+        self.builder.new_output_connection(connection)
+    }
+
+    /// Creates an operator implementation from supplied logic constructor.
+    pub fn build<B, L>(self, constructor: B)
+    where
+        B: FnOnce(Vec<Capability<G::Timestamp>>) -> L,
+        L: Future + 'static,
+    {
+        let operator_waker = self.operator_waker;
+        let registered_wakers = self.registered_wakers;
+        let shared_frontiers = self.shared_frontiers;
+        let input_handles = self.input_handles;
+        self.builder.build_reschedule(move |caps| {
+            let mut logic_fut = Some(Box::pin(constructor(caps)));
+            move |new_frontiers| {
+                // First, discover if there are any frontier notifications
+                {
+                    let mut old_frontiers = shared_frontiers.borrow_mut();
+                    for ((old, pending), new) in old_frontiers.iter_mut().zip(new_frontiers) {
+                        if PartialOrder::less_than(&old.borrow(), &new.frontier()) {
+                            *old = new.frontier().to_owned();
+                            *pending = true;
+                        }
+                    }
+                }
+                // Then, wake up any registered Wakers for the input streams while taking care to
+                // not reactivate the timely operator since that would lead to an infinite loop.
+                // This ensures that the input handles wake up properly when managed by other
+                // executors, e.g a `select!` that provides its own Waker implementation.
+                {
+                    operator_waker.active.store(true, Ordering::SeqCst);
+                    let mut registered_wakers = registered_wakers.borrow_mut();
+                    for waker in registered_wakers.drain(..) {
+                        waker.wake();
+                    }
+                    operator_waker.active.store(false, Ordering::SeqCst);
+                }
+                // Schedule the logic future if any of the wakers above marked the task as ready
+                if let Some(fut) = logic_fut.as_mut() {
+                    if operator_waker.task_ready.load(Ordering::SeqCst) {
+                        let waker = futures_util::task::waker_ref(&operator_waker);
+                        let mut cx = Context::from_waker(&waker);
+                        if Pin::new(fut).poll(&mut cx).is_ready() {
+                            // We're done with logic so deallocate the task
+                            logic_fut = None;
+                        }
+                        operator_waker.task_ready.store(false, Ordering::SeqCst);
+                    }
+                }
+                // The timely operator needs to be kept alive if the task is pending or if there
+                // are live input handles, perhaps in an another executor
```

Comment:
if the logic future is dead, then what happens if their are live input handles? nothing will be able to read them, right?

Reply:
It was supposed to support a usecase where you've sent the handles to another executor but the original operator has terminated. I removed the `input_handle` logic and now just panic if a handle is used after the operator has died